File: framec_tests/src/event_handler.rs
//! Test event handler parameters, local variables, and return values.

type Log = Vec<String>;
include!(concat!(env!("OUT_DIR"), "/", "event_handler.rs"));

impl EventHandler {
    pub fn log(&mut self, msg: String, val: i32) {
        self.tape.push(format!("{}={}", msg, val));
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn single_parameter() {
        let mut sm = EventHandler::new();
        sm.log_it(2);
        assert_eq!(sm.tape, vec!["x=2"]);
    }

    #[test]
    fn compute_two_parameters() {
        let mut sm = EventHandler::new();
        sm.log_add(-3, 10);
        assert_eq!(sm.tape, vec!["a=-3", "b=10", "a+b=7"]);
    }

    #[test]
    fn return_local_variable() {
        let mut sm = EventHandler::new();
        let ret = sm.log_return(13, 21);
        assert_eq!(sm.tape, vec!["a=13", "b=21", "r=34"]);
        assert_eq!(ret, 34);
    }

    #[test]
    fn pass_result() {
        let mut sm = EventHandler::new();
        sm.pass_add(5, -12);
        assert_eq!(sm.tape, vec!["p=-7"]);
    }

    #[test]
    fn pass_and_return_result() {
        let mut sm = EventHandler::new();
        let ret = sm.pass_return(101, -59);
        assert_eq!(sm.tape, vec!["r=42", "p=42"]);
        assert_eq!(ret, 42);
    }
}

File: framec_tests_javascript/test/event_handler.js
const { describe, it } = require("mocha");
const assert = require("assert");

const EventHandler = require("../output/event_handler");

class EventHandlerController extends EventHandler {

	constructor() {
	  super()
	}
	log_do(msg, val) {
		this.tape.push(`${msg}=${val}`)
	}
};

describe("Event handler", ()=>{

	it("Single parameter", () => {
		let sm = new EventHandlerController();
		sm.LogIt(2);
		assert.deepStrictEqual(sm.tape, ["x=2"])

	})

	it("Compute two parameters", ()=>{
		let sm = new EventHandlerController()
		sm.LogAdd(-3, 10)
		assert.deepStrictEqual(sm.tape, ["a=-3", "b=10", "a+b=7"]);
	})
	
	it("Return local variable", ()=> {
		let sm = new EventHandlerController()
		let ret = sm.LogReturn(13, 21);
		assert.deepStrictEqual(sm.tape, ["a=13", "b=21", "r=34"]);
		assert.deepStrictEqual(ret, 34);
	})

	it("Pass result", ()=>{
		let sm = new EventHandlerController()
		sm.PassAdd(5, -12);
		assert.deepStrictEqual(sm.tape, ["p=-7"])
	})

	it("Pass and return result", ()=> {
		let sm = new EventHandlerController()
        let ret = sm.PassReturn(101, -59);
        assert.deepStrictEqual(sm.tape, ["r=42", "p=42"]);
        assert.deepStrictEqual(ret, 42);
	})
})